[SPARK-24441][SS] Expose total estimated size of states in HDFSBackedStateStoreProvider

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

This patch exposes the estimation of size of cache (loadedMaps) in HDFSBackedStateStoreProvider as a custom metric of StateStore.

The rationalize of the patch is that state backed by HDFSBackedStateStoreProvider will consume more memory than the number what we can get from query status due to caching multiple versions of states. The memory footprint to be much larger than query status reports in situations where the state store is getting a lot of updates: while shallow-copying map incurs additional small memory usages due to the size of map entities and references, but row objects will still be shared across the versions. If there're lots of updates between batches, less row objects will be shared and more row objects will exist in memory consuming much memory then what we expect. 

While HDFSBackedStateStore refers loadedMaps in HDFSBackedStateStoreProvider directly, there would be only one `StateStoreWriter` which refers a StateStoreProvider, so the value is not exposed as well as being aggregated multiple times. Current state metrics are safe to aggregate for the same reason.

## How was this patch tested?

Tested manually. Below is the snapshot of UI page which is reflected by the patch: 

<img width="601" alt="screen shot 2018-06-05 at 10 16 16 pm" src="https://user-images.githubusercontent.com/1317309/40978481-b46ad324-690e-11e8-9b0f-e80528612a62.png">

Please refer "estimated size of states cache in provider total" as well as "count of versions in state cache in provider".

HeartSaVioR committed:
[SPARK-24441][SS] Expose total size of states in HDFSBackedStateStoreProvider

* NOTE: This commit is squashed version of commits which are already reviewed by others
* expose estimation of size of cache (loadMaps) in HDFSBackedStateStoreProvider
  * as a custom metric of StateStore
* lowercase description of new metric: existing metrics are all lowercase
* reduce indentation level via starting with synchronized from the start of func. definition
* expose another custom metric: (average) count of versions stored in loadedMaps
* Also expose custom metrics in state store to StreamingQueryStatus as well
  * average metric is implemented a bit tricky way, so just exclude average metrics for now
* rename metricProviderLoaderMapSize to metricProviderLoaderMapSizeBytes
* Fix broken tests
  * scala Map doesn't look like working well with json4s: followed up the existing approach to resolve
* Add more assertion

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #91347 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91347/testReport)** for PR 21469 at commit [`dc11338`](https://github.com/apache/spark/commit/dc11338650842a246a4bce9280d130607ceca281).

SparkQA commented:
**[Test build #91347 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91347/testReport)** for PR 21469 at commit [`dc11338`](https://github.com/apache/spark/commit/dc11338650842a246a4bce9280d130607ceca281).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91347/
Test PASSed.

SparkQA commented:
**[Test build #91375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91375/testReport)** for PR 21469 at commit [`6c1c30b`](https://github.com/apache/spark/commit/6c1c30b9a2722cfd1eb8ee72575f142c00365fb3).

SparkQA commented:
**[Test build #91375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91375/testReport)** for PR 21469 at commit [`6c1c30b`](https://github.com/apache/spark/commit/6c1c30b9a2722cfd1eb8ee72575f142c00365fb3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91375/
Test PASSed.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:
```diff
@@ -181,6 +182,12 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
     }
   }
 
+  def getCustomMetricsForProvider(): Map[StateStoreCustomMetric, Long] = {
```
tiny nit:

```scala
def getCustomMetricsForProvider(): Map[StateStoreCustomMetric, Long] = synchronized {
  Map(metricProviderLoaderMapSize -> SizeEstimator.estimate(loadedMaps))
}
```

HyukjinKwon commented:
Shall we make the PR title complete? Looks truncated.

HeartSaVioR commented:
Thanks @HyukjinKwon for reviewing. Addressed PR title as well as fixing nit.

SparkQA commented:
**[Test build #91382 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91382/testReport)** for PR 21469 at commit [`933fb2e`](https://github.com/apache/spark/commit/933fb2e329b7e76ae5087e6ce58dd3359c78d5f1).

jose-torres commented:
LGTM. To clarify the description, we expect the memory footprint to be much larger than query status reports in situations where the state store is getting a lot of updates?

HeartSaVioR commented:
@jose-torres 
Ah yes I forgot that shallow copy has been occurring, so while new map should hold necessary size of map entries but row object will be shared across versions. Thanks for pointing it out. Will update the description.

arunmahadevan commented:
@HeartSaVioR , may be then this should be reported in the "memoryUsedBytes" in the StateOperatorProgress (value reported in StreamingQueryProgress) because currently the usage reported does not reflect the memory used for the cache.

Question: in the screenshot "Estimated size of states cache in provider total" is 3.3 MB whereas the "memory used by state total" is 20.6 KB with "total number of state rows" = 2. This 150x difference is expected with just 2 rows in the state? Were there 100 versions of the map in the sample output you posted?

HeartSaVioR commented:
@arunmahadevan 
I didn't add the metric to StateOperatorProgress cause this behavior is specific to HDFSBackedStateStoreProvider (though this is only one implementation available in Apache Spark) so not sure this metric can be treated as a general one. (@tdas what do you think about this?)

Btw, the cache is going to clean up when maintenance operation is in progress, so there could be more than 100 versions in map. Not sure why it shows 150x, but I couldn't find missing spot on the patch. Maybe the issue is from SizeEstimator.estimate()?

One thing we need to check is how SizeEstimator.estimate() calculate the memory usage when Unsafe row objects are shared across versions. If SizeEstimator adds the size of object whenever it is referenced, it will report much higher memory usage than actual.

HeartSaVioR commented:
Also added custom metric for the count of versions stored in loadedMaps.

This is a new screenshot:
<img width="601" alt="screen shot 2018-06-05 at 10 16 16 pm" src="https://user-images.githubusercontent.com/1317309/40978481-b46ad324-690e-11e8-9b0f-e80528612a62.png">

SparkQA commented:
**[Test build #91486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91486/testReport)** for PR 21469 at commit [`345397d`](https://github.com/apache/spark/commit/345397dd530898697ef338ec55b81ad11ece4dc0).

arunmahadevan commented:
>I didn't add the metric to StateOperatorProgress cause this behavior is specific to HDFSBackedStateStoreProvider

May be this can be reported as a custom metrics and keep it optional and that way its not tied to any specific implementation.

SparkQA commented:
**[Test build #91486 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91486/testReport)** for PR 21469 at commit [`345397d`](https://github.com/apache/spark/commit/345397dd530898697ef338ec55b81ad11ece4dc0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class StateStoreCustomAverageMetric(name: String, desc: String) extends StateStoreCustomMetric`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91486/
Test PASSed.

HyukjinKwon commented:
@jose-torres is it good to go?

SparkQA commented:
**[Test build #91503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91503/testReport)** for PR 21469 at commit [`af57f26`](https://github.com/apache/spark/commit/af57f26239ef635cb9837e9ebcf37fbdaa215480).

HeartSaVioR commented:
@arunmahadevan 
Added custom metrics in state store to streaming query status as well. You can see `providerLoadedMapSize` is added to `stateOperators.customMetrics` in below output.

I have to exclude `providerLoadedMapCountOfVersions` from the list, since average metric is implemented a bit tricky and doesn't look like easy to aggregate for streaming query status. 
We may want to reimplement SQLMetric and subclasses to make sure everything works correctly without any tricky approach, but that doesn't look like trivial to address and I think this is out of scope on this PR.

```
18/06/06 22:51:23 INFO MicroBatchExecution: Streaming query made progress: {
  "id" : "7564a0b7-e3b2-4d53-b246-b774ab04e586",
  "runId" : "8dd34784-080c-4f86-afaf-ac089902252d",
  "name" : null,
  "timestamp" : "2018-06-06T13:51:15.467Z",
  "batchId" : 4,
  "numInputRows" : 547,
  "inputRowsPerSecond" : 67.15776550030694,
  "processedRowsPerSecond" : 65.94333936106088,
  "durationMs" : {
    "addBatch" : 7944,
    "getBatch" : 1,
    "getEndOffset" : 0,
    "queryPlanning" : 61,
    "setOffsetRange" : 5,
    "triggerExecution" : 8295,
    "walCommit" : 158
  },
  "eventTime" : {
    "avg" : "2018-06-06T13:51:10.313Z",
    "max" : "2018-06-06T13:51:14.250Z",
    "min" : "2018-06-06T13:51:07.098Z",
    "watermark" : "2018-06-06T13:50:36.676Z"
  },
  "stateOperators" : [ {
    "numRowsTotal" : 20,
    "numRowsUpdated" : 16,
    "memoryUsedBytes" : 26679,
    "customMetrics" : {
      "providerLoadedMapSize" : 181911
    }
  } ],
  "sources" : [ {
    "description" : "KafkaV2[Subscribe[apachelogs-v2]]",
    "startOffset" : {
      "apachelogs-v2" : {
        "2" : 489056,
        "4" : 489053,
        "1" : 489055,
        "3" : 489051,
        "0" : 489053
      }
    },
    "endOffset" : {
      "apachelogs-v2" : {
        "2" : 489056,
        "4" : 489053,
        "1" : 489055,
        "3" : 489051,
        "0" : 489053
      }
    },
    "numInputRows" : 547,
    "inputRowsPerSecond" : 67.15776550030694,
    "processedRowsPerSecond" : 65.94333936106088
  } ],
  "sink" : {
    "description" : "org.apache.spark.sql.execution.streaming.ConsoleSinkProvider@60999714"
  }
}
```

arunmahadevan commented:
Nice, LGTM.

SparkQA commented:
**[Test build #91503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91503/testReport)** for PR 21469 at commit [`af57f26`](https://github.com/apache/spark/commit/af57f26239ef635cb9837e9ebcf37fbdaa215480).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91503/
Test FAILed.

SparkQA commented:
**[Test build #91509 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91509/testReport)** for PR 21469 at commit [`7ec3242`](https://github.com/apache/spark/commit/7ec32427cf0fda82d2f936fa0aab62e274a8c034).

SparkQA commented:
**[Test build #91509 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91509/testReport)** for PR 21469 at commit [`7ec3242`](https://github.com/apache/spark/commit/7ec32427cf0fda82d2f936fa0aab62e274a8c034).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91509/
Test PASSed.

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala:
```diff
@@ -231,7 +231,7 @@ class StreamingQueryListenerSuite extends StreamTest with BeforeAndAfter {
     val listener = new EventCollector
     withListenerAdded(listener) {
-      for (i <- 1 to 100) {
+      for (i <- 1 to 50) {
```
After the patch this test starts failing: it just means there's more time needed to run this loop 100 times, and doesn't mean the logic is broken. Decreasing number works for me.

jose-torres replied:
Makes sense, and I agree with the implicit claim that this slowdown isn't too worrying.

SparkQA commented:
**[Test build #91523 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91523/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91523/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #91526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91526/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).

SparkQA commented:
**[Test build #91526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91526/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91526/
Test FAILed.

HeartSaVioR commented:
retest this please

SparkQA commented:
**[Test build #91535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91535/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).

SparkQA commented:
**[Test build #91535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91535/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91535/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #91596 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91596/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).

SparkQA commented:
**[Test build #91596 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91596/testReport)** for PR 21469 at commit [`3c80cad`](https://github.com/apache/spark/commit/3c80cad32c056a24a7f5ffd7ab0ae3f7e096a62d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91596/
Test PASSed.

jose-torres commented:
(Sorry to comment after so long with such a minor change - I've been busy with spark summit)

metricProviderLoaderMapSize should be metricProviderLoaderMapSizeBytes, both for clarity and consistency with memoryUsedBytes. Other than that LGTM.

SparkQA commented:
**[Test build #91625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91625/testReport)** for PR 21469 at commit [`62fc395`](https://github.com/apache/spark/commit/62fc395d29b67a827612e4e7f4bf134f5817a33c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91625/
Test PASSed.

arunmahadevan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:
```diff
@@ -247,6 +253,14 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
+    StateStoreCustomSizeMetric("providerLoadedMapSizeBytes",
+      "estimated size of states cache in provider")
+
+  private lazy val metricProviderLoaderCountOfVersionsInMap: StateStoreCustomAverageMetric =
```
Why is "metricProviderLoaderCountOfVersionsInMap" an average metrics? The other metrics like "numTotalStateRows" and even "providerLoadedMapSizeBytes" is count metric. Shouldn't this be similar?

HeartSaVioR replied:
This should be average to show `min, med, max` in SQL metrics UI, as I pasted capture of UI before. Summing them all doesn't give meaningful value.

arunmahadevan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:
```diff
@@ -112,14 +122,19 @@ trait StateStoreWriter extends StatefulOperator { self: SparkPlan =>
-      longMetric(metric.name) += value
+    storeMetrics.customMetrics.foreach {
+      case (metric: StateStoreCustomAverageMetric, value) =>
+        longMetric(metric.name).set(value * 1.0d)
```
How does this get accumulated ? It seems the value last set may get propagated to the driver.

HeartSaVioR replied:
If my understanding is right, the metric object (return of `longMetric()`) is different between each task, so the object will be different for each batch and each task. (TaskMetric is serialized and deserialized so it can't be shared between tasks.)

And actually the metric values are not aggregated into an SQLMetric object. The values are just aggregated and represented in SQLAppStatusListener.

https://github.com/apache/spark/blob/f5af86ea753c446df59a0a8c16c685224690d633/sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala#L162-L174

https://github.com/apache/spark/blob/f5af86ea753c446df59a0a8c16c685224690d633/sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala#L147-L160

<img width="489" alt="screen shot 2018-06-12 at 9 17 14 am" src="https://user-images.githubusercontent.com/1317309/41263432-024efe4a-6e22-11e8-92f9-24d1f73776a9.png">

HeartSaVioR replied:
We would be better to think about the actual benefit of exposing the value, rather than how to expose the value to somewhere. If we define it as count and do aggregation as summation, the aggregated value will be `(partition count * versions)` which might be hard for end users to find the meaning from the value.

I'm afraid that exposing this to StreamingQuery as average is not trivial, especially SQLMetric is defined as `AccumulatorV2[Long, Long]` so only single Long value can be passed. Under the restriction, we couldn't define `merge` operation for `average metric`.

HeartSaVioR commented:
Kindly ping again to @tdas

HeartSaVioR commented:
adding cc. to @zsxwing since he has been reviewing PRs for SS so far.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #92329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92329/testReport)** for PR 21469 at commit [`62fc395`](https://github.com/apache/spark/commit/62fc395d29b67a827612e4e7f4bf134f5817a33c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92329/
Test PASSed.

HeartSaVioR committed:
Add two more metrics: state cache hit / miss count in HDFS state provider

SparkQA commented:
**[Test build #92542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92542/testReport)** for PR 21469 at commit [`96115bc`](https://github.com/apache/spark/commit/96115bcc48cfe67c4b7bd37315963b9ba1366b3a).

HeartSaVioR commented:
Rebased to fix conflict, and added new commit (last one: c9aada5) to represent cache hit / miss count in HDFS state provider. This is actually helpful for SPARK-24717 to determine proper value of configuration, but with this commit SPARK-24717 should be on top of this PR, so just added it here to avoid rebase hell.

SparkQA commented:
**[Test build #92544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92544/testReport)** for PR 21469 at commit [`c9aada5`](https://github.com/apache/spark/commit/c9aada520889b87ace0886805910f0d56d099bd2).

SparkQA commented:
**[Test build #92542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92542/testReport)** for PR 21469 at commit [`96115bc`](https://github.com/apache/spark/commit/96115bcc48cfe67c4b7bd37315963b9ba1366b3a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class StateStoreCustomSumMetric(name: String, desc: String) extends StateStoreCustomMetric`

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92542/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #92544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92544/testReport)** for PR 21469 at commit [`c9aada5`](https://github.com/apache/spark/commit/c9aada520889b87ace0886805910f0d56d099bd2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class StateStoreCustomSumMetric(name: String, desc: String) extends StateStoreCustomMetric`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92544/
Test PASSed.

SparkQA commented:
**[Test build #92551 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92551/testReport)** for PR 21469 at commit [`c9aada5`](https://github.com/apache/spark/commit/c9aada520889b87ace0886805910f0d56d099bd2).

SparkQA commented:
**[Test build #92551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92551/testReport)** for PR 21469 at commit [`c9aada5`](https://github.com/apache/spark/commit/c9aada520889b87ace0886805910f0d56d099bd2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class StateStoreCustomSumMetric(name: String, desc: String) extends StateStoreCustomMetric`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92551/
Test PASSed.

HeartSaVioR commented:
Now I'm thinking about removing "metricProviderLoaderCountOfVersionsInMap" and also removing StateStoreCustomAverageMetric, since the value doesn't look correct with stream-stream join which handles multiple states in an operation.

We may try out SQLMetric to extend AccumulatorV2[(kind of Metric class), (kind of Metric class)] which Metric class can handle merging multiple values correctly according to the kind of metric, but this is beyond the scope of PR, so just removing "metricProviderLoaderCountOfVersionsInMap" would be clearer for this patch for now.

What do you think, @arunmahadevan @jose-torres ?

jose-torres commented:
Sure, I don't mind if we remove that metric.

HeartSaVioR committed:
Get rid of metricProviderLoaderCountOfVersionsInMap and also StateStoreCustomAverageMetric

SparkQA commented:
**[Test build #92839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92839/testReport)** for PR 21469 at commit [`c4a6d11`](https://github.com/apache/spark/commit/c4a6d11e230759a0d068024cba04c31a0e7903a9).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92839/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #92860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92860/testReport)** for PR 21469 at commit [`c4a6d11`](https://github.com/apache/spark/commit/c4a6d11e230759a0d068024cba04c31a0e7903a9).

SparkQA commented:
**[Test build #92860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92860/testReport)** for PR 21469 at commit [`c4a6d11`](https://github.com/apache/spark/commit/c4a6d11e230759a0d068024cba04c31a0e7903a9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92860/
Test PASSed.

HeartSaVioR committed:
Fix line which broke Spark Scala Style guide

HeartSaVioR committed:
Refine test code a bit to remove duplicated code

HeartSaVioR committed:
Refine a bit more

SparkQA commented:
**[Test build #93224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93224/testReport)** for PR 21469 at commit [`5b203d4`](https://github.com/apache/spark/commit/5b203d4967eda3a09f7c8d83cf86e7ac6a427182).

SparkQA commented:
**[Test build #93224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93224/testReport)** for PR 21469 at commit [`5b203d4`](https://github.com/apache/spark/commit/5b203d4967eda3a09f7c8d83cf86e7ac6a427182).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93224/
Test PASSed.

HeartSaVioR committed:
Refine a little more

SparkQA commented:
**[Test build #93257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93257/testReport)** for PR 21469 at commit [`32d0418`](https://github.com/apache/spark/commit/32d041878b7dcd20794250853063dab4bac09118).

SparkQA commented:
**[Test build #93257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93257/testReport)** for PR 21469 at commit [`32d0418`](https://github.com/apache/spark/commit/32d041878b7dcd20794250853063dab4bac09118).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93257/
Test PASSed.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala:
```diff
@@ -48,12 +49,24 @@ class StateOperatorProgress private[sql](
+      ("numRowsUpdated" -> JInt(numRowsUpdated)) ~
+      ("memoryUsedBytes" -> JInt(memoryUsedBytes))
+
+    if (!customMetrics.isEmpty) {
```
You are already handling the case of map being empty in `safeMapToJValue` by adding JNothing. Doesnt JNothing values just get dropped from the json text any way?

HeartSaVioR replied:
Actually didn't notice that. Thanks for letting me know! Will simplify.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala:
```diff
@@ -48,12 +49,24 @@ class StateOperatorProgress private[sql](
-    ("numRowsTotal" -> JInt(numRowsTotal)) ~
-    ("numRowsUpdated" -> JInt(numRowsUpdated)) ~
-    ("memoryUsedBytes" -> JInt(memoryUsedBytes))
+    def safeMapToJValue[T](map: ju.Map[String, T], valueToJValue: T => JValue): JValue = {
```
T is always Long. Why make a generic function for that? This does not even need a separate function.

HeartSaVioR replied:
I've first trying to leverage `StreamingQueryProgress.safeMapToJValue` but can't find proper place to move to be co-used, so I simply copied it. Will simplify the code block and inline.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala:
```diff
@@ -81,10 +81,10 @@ class SQLMetric(val metricType: String, initValue: Long = 0L) extends Accumulato
+  val SUM_METRIC = "sum"
+  val SIZE_METRIC = "size"
+  val TIMING_METRIC = "timing"
+  val AVERAGE_METRIC = "average"
```
why this change?

HeartSaVioR replied:
It was to handle exception case while aggregating custom metrics, especially filtering out average since it is not aggregated correctly. Since we remove custom average metric, we no longer need to filter out them. Will revert the change as well as relevant logic.

tdas reviewed (changes requested):
This looks good. Only two comments.

HeartSaVioR committed:
Address review comments from @tdas

HeartSaVioR commented:
@tdas Thanks for the review! Addressed review comments.

SparkQA commented:
**[Test build #93869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93869/testReport)** for PR 21469 at commit [`ed072fc`](https://github.com/apache/spark/commit/ed072fcf057f982275d0daf69787ed812f03e87b).

HeartSaVioR commented:
Retest this, please

SparkQA commented:
**[Test build #93906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93906/testReport)** for PR 21469 at commit [`ed072fc`](https://github.com/apache/spark/commit/ed072fcf057f982275d0daf69787ed812f03e87b).

SparkQA commented:
**[Test build #93906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93906/testReport)** for PR 21469 at commit [`ed072fc`](https://github.com/apache/spark/commit/ed072fcf057f982275d0daf69787ed812f03e87b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93906/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #93927 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93927/testReport)** for PR 21469 at commit [`ed072fc`](https://github.com/apache/spark/commit/ed072fcf057f982275d0daf69787ed812f03e87b).

SparkQA commented:
**[Test build #93927 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93927/testReport)** for PR 21469 at commit [`ed072fc`](https://github.com/apache/spark/commit/ed072fcf057f982275d0daf69787ed812f03e87b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93927/
Test PASSed.

HeartSaVioR commented:
@tdas Kindly reminder.

tdas commented:
I am having a second thoughts about this. Exposing the entire memory usage of all the loaded maps as another custom metric .... just adds more confusion. Rather the point of the the main state metric `memoryUsedBytes` is to capture how much memory is occupied because of the one partition of the state, and that implicitly should cover all the loaded versions of that state partition. So I strongly feel that instead of adding a custom metric, we should change the existing `memoryUsedBytes` to capture all the memory. 

I am fine adding the custom metrics hit and miss counts. No questions about that. 

What do you think?

HeartSaVioR commented:
My series of patches could be possible based on two metrics: `size for memory usage of latest version` and `size for total memory usage of loaded versions`. SPARK-24717 (#21700) enabled the possibility to tune the overall state memory usage, and if end users have either one metric they couldn't tune it.

IMHO, I'm not 100% sure how much this patch provides confusion to the end users, but if the intention of `memoryUsedBytes` is for measuring overall state partition, what about replacing `memoryUsedBytes` as `size for total memory usage of loaded versions`, but also placing `size for memory usage of latest version` to custom metric?

tdas commented:
@HeartSaVioR I think I agree with a second approach that you suggested. So 
`memoryUsedBytes` => `size for total memory usage of loaded versions` and
`customMetric` => `size for memory usage of latest version`

HeartSaVioR committed:
Use memoryUsedBytes to total memory usage of loaded versions

customMetric.stateOnCurrentVersionSizeBytes to size for memory usage of current version

HeartSaVioR commented:
@tdas Thanks for the feedback! Updated the PR.

SparkQA commented:
**[Test build #95012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95012/testReport)** for PR 21469 at commit [`545081e`](https://github.com/apache/spark/commit/545081e4b7a9bb37442f7d558bc28db790742610).

tdas commented:
LGTM.

tdas commented:
Merged to master.

HeartSaVioR commented:
Thanks all for reviewing and thanks @tdas for merging this in!

tdas commented:
Unfortunately this PR broke the master build. Looks like some import that probably got removed in the other PR I merged, which didnt create any direct conflict.

HeartSaVioR commented:
@tdas In case of you are not working on the patch, I'm working on the fix and will provide minor PR.

tdas commented:
I did. Fixed the import
